src/backend/distributed/utils/resource_lock.c

```diff
@@ -1060,3 +1086,218 @@ CitusLockTableAclCheck(Oid relationId, LOCKMODE lockmode, Oid userId)
+			/* if local node is one of the targets, acquire the lock locally */
+			if (workerNode->groupId == localGroupId)
+			{
+				LockRelationIfExists(
+					cstring_to_text(qualifiedRelationName),
+					lockModeCString,
+					nowait);
+
+				continue;
+			}
+
+			SendCommandToWorker(nodeName, nodePort, lockRelationCommand->data);
+		}
+	}
+}
+
+
+/*
+ * AcquireDistributedLockOnRelations filters relations before passing them to
+ * AcquireDistributedLockOnRelations_Internal to acquire the locks.
+ *
+ * Only tables, views, and foreign tables can be locked with this function. Other relations
+ * will cause an error.
+ *
+ * Gracefully locks relations so no errors are thrown for things like invalid id-s
+ * or missing relations on worker nodes.
+ *
+ * Considers different types of relations based on a 'configs' parameter:
+ * - DIST_LOCK_DEFAULT: locks citus tables
+ * - DIST_LOCK_VIEWS_RECUR: locks citus tables that locked views are dependent on recursively
+ * - DIST_LOCK_REFERENCING_TABLES: locks tables that refer to locked citus tables with a foreign key
+ * - DIST_LOCK_NOWAIT: throws an error if the lock is not immediately available
+ */
+void
+AcquireDistributedLockOnRelations(List *relationList, LOCKMODE lockMode, uint32 configs)
+{
+	/* nothing to do if there is no metadata at worker nodes */
+	if (!ClusterHasKnownMetadataWorkers())
+	{
+		return;
+	}
+
+	List *distributedRelationList = NIL;
+
+	RangeVar *rangeVar = NULL;
+	List *relations = list_copy(relationList);
+	foreach_ptr(rangeVar, relations)
+	{
+		Oid relationId = RangeVarGetRelid(rangeVar, NoLock, false);
+
+		if ((configs & DIST_LOCK_VIEWS_RECUR) > 0 && get_rel_relkind(relationId) ==
+			RELKIND_VIEW)
+		{
+			ObjectAddress viewAddress = { 0 };
+			Oid schemaOid = RangeVarGetCreationNamespace(rangeVar);
+			ObjectAddressSet(viewAddress, schemaOid, relationId);
+
+			List *distDependencies = GetDistributableDependenciesForObject(&viewAddress);
+
+			ObjectAddress *address = NULL;
+			foreach_ptr(address, distDependencies)
+			{
+				distributedRelationList = list_append_unique_oid(distributedRelationList,
+																 address->objectId);
+			}
+
+			continue;
+		}
+
+		if (!IsCitusTable(relationId))
+		{
+			continue;
+		}
+
+		if (list_member_oid(distributedRelationList, relationId))
+		{
+			continue;
+		}
+
+		distributedRelationList = lappend_oid(distributedRelationList, relationId);
+
+		if ((configs & DIST_LOCK_REFERENCING_TABLES) > 0)
+		{
+			Oid referencingRelationId = InvalidOid;
+			CitusTableCacheEntry *cacheEntry = GetCitusTableCacheEntry(relationId);
+			Assert(cacheEntry != NULL);
+
+			List *referencingTableList = cacheEntry->referencingRelationsViaForeignKey;
+			foreach_oid(referencingRelationId, referencingTableList)
+			{
+				distributedRelationList = list_append_unique_oid(distributedRelationList,
+																 referencingRelationId);
+			}
+		}
+	}
+
+	if (distributedRelationList != NIL)
+	{
+		if (!IsCoordinator() && !CoordinatorAddedAsWorkerNode())
```

Comment:
As we discussed, this is the right thing to do. HOWEVER, this error means we are breaking existing users, especially the ones that are NOT using coordinator for running `TRUNCATE` from the coordinator. 

So, one idea could be to add a GUC such as `citus.prevent_locking_when_coordinator_not_added` (this is a terrible name, we need a better one :)), which is by default throwing this error. But, users could change the behavior. 

I think it'd be good to get @marcocitus 's thoughts on this.